### PR TITLE
Flywheel Implementation

### PR DESCRIPTION
**Pull request by Urbanmass**
hopefully everything works

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/shooter/Flywheel.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import frc.robot.Constants.HardwareConstants;
+
+public class Flywheel extends SubsystemBase {
+  private final FlywheelIO io;
+  private final FlywheelIOInputsAutoLogged inputs = new FlywheelIOInputsAutoLogged();
+  private final SimpleMotorFeedforward ffModel;
+  
+  public Flywheel(FlywheelIO io) {
+    this.io = io;
+    switch (HardwareConstants.currentMode) {
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/shooter/Flywheel.java:34]: switch case with enums is very nice

File: src/main/java/frc/robot/subsystems/shooter/Flywheel.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      case REAL:
+      case REPLAY:
+        ffModel = new SimpleMotorFeedforward(0.1, 0.05);
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/shooter/Flywheel.java:37]: should these values be constants? Once again, I'm not too familiar with the sim stuff

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+package frc.robot.subsystems.shooter;
+
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.simulation.FlywheelSim;
+
+public class FlywheelIOSim implements FlywheelIO {
+  private FlywheelSim sim = new FlywheelSim(DCMotor.getNEO(1), 1.5, 0.004);
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:22]: We're using neos????

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  private boolean closedLoop = false; //closed loop is pid
+  private double ffVolts = 0.0;
+  private double appliedVolts = 0.0;
+
+  @Override
+  public void updateInputs(FlywheelIOInputs inputs) {
+    if (closedLoop) {
+      appliedVolts =
+          MathUtil.clamp(pid.calculate(sim.getAngularVelocityRadPerSec()) + ffVolts, -12.0, 12.0);
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:33]: Make constants -- seems like a good thing to put in HardwareConstants (if we still have that)

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  public void setVoltage(double volts) {
+    closedLoop = false;
+    appliedVolts = volts;
+    sim.setInputVoltage(volts);
+  }
+
+  @Override
+  public void setVelocity(double velocityRPM, double ffVolts) {
+    closedLoop = true;
+    pid.setSetpoint(velocityRPM / 60.0);
```
Comment (JacksonElia) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:55]: Make a Constant

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java (hunk 3 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:22]: no, i believe these should be krakens

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java (hunk 4 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:33]: what should be constants?

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java (hunk 5 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:55]: for what

File: src/main/java/frc/robot/subsystems/shooter/Flywheel.java (hunk 1 above)
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/Flywheel.java:34]: i'd prefer if we did all the pid/ff  tuning within the implementations rather than through this switch statement.

File: src/main/java/frc/robot/subsystems/shooter/Flywheel.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        ffModel = new SimpleMotorFeedforward(0.0, 0.0);
+        break;
+    }
+  }
+
+  public void setFlywheelVoltage(double desiredVoltage) {
+    io.setVoltage(desiredVoltage);
+  }
+
+  public void setFlywheelVelocity(double desiredRPM, double ffVolts) {
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/Flywheel.java:54]: If you set up pid/ff correctly you should only have to set the desiredRPM, you shouldn't have to have a separate parameter for ffVolts

File: src/main/java/frc/robot/subsystems/shooter/Flywheel.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    io.setVelocity(desiredRPM, ffVolts);
+  }
+
+  @Override
+  public void periodic() {
+    // This method will be called once per scheduler run
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/Flywheel.java:60]: you need to run `io.updateInputs()` and `Logger.processInputs()` wthin the periodic method

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.wpilibj.simulation.FlywheelSim;
+
+public class FlywheelIOSim implements FlywheelIO {
+  private FlywheelSim sim = new FlywheelSim(DCMotor.getNEO(1), 1.5, 0.004);
+  private PIDController pid = new PIDController(0.0, 0.0, 0.0);
+
+  private boolean closedLoop = false; //closed loop is pid
+  private double ffVolts = 0.0;
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:26]: For sim i'd just create a SimpleMotorFeedforward

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIO.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+// GNU General Public License for more details.
+
+package frc.robot.subsystems.shooter;
+
+import org.littletonrobotics.junction.AutoLog;
+
+public interface FlywheelIO {
+  @AutoLog
+  public static class FlywheelIOInputs { //variables that 
+    public double positionRotaions = 0.0;  //positions in radians | convert to rpms
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIO.java:21]: spelled rotations incorrectly

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  private double appliedVolts = 0.0;
+
+  @Override
+  public void updateInputs(FlywheelIOInputs inputs) {
+    if (closedLoop) {
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:31]: personally I wouldn't do this `if (closedLoop)` thingy (especially since I doubt we'll use openLoop on the flywheel) and just setup the pid within a method like `setVelocity()`

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIO.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  public default void setVoltage(double volts) {}
+
+  /** Run closed loop at the specified velocity. */
+  public default void setVelocity(double velocityRPM, double ffVolts) {}
+
+  /** Stop in open loop. */
+  public default void stop() {}
+
+  /** Set velocity PID constants. */
+  public default void configurePID(double kP, double kI, double kD) {}
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIO.java:40]: Right now, I don't really like this, I'd prefer you just tune it with constants within the implementation.

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+import com.ctre.phoenix6.controls.Follower;
+import com.ctre.phoenix6.controls.VelocityVoltage;
+import com.ctre.phoenix6.controls.VoltageOut;
+import com.ctre.phoenix6.hardware.TalonFX;
+import com.ctre.phoenix6.signals.InvertedValue;
+import com.ctre.phoenix6.signals.NeutralModeValue;
+import frc.robot.Constants.FlywheelConstants;
+
+public class FlywheelIOTalonFX implements FlywheelIO {
+  private static final double GEAR_RATIO = FlywheelConstants.GEAR_RATIO; //Gear ratio of two motors on flywheel
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java:29]: gearRatio. Also why is this here?

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  public void setVoltage(double volts) { //some method to set voltage for motor
+    leaderFlywheelMotor.setControl(new VoltageOut(volts));
+  }
+
+  @Override
+  public void setVelocity(double velocityRPM, double ffVolts) {//updating voltage to alter velocity of the motor
+    leaderFlywheelMotor.setControl(
+        new VelocityVoltage(
+            velocityRPM / 60.0,
+            0.0,
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java:80]: u don't need all the add-ons for VelocityVoltage, just the first parameter should be enough.

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+  @Override
+  public void setVoltage(double volts) { //some method to set voltage for motor
+    leaderFlywheelMotor.setControl(new VoltageOut(volts));
+  }
+
+  @Override
+  public void setVelocity(double velocityRPM, double ffVolts) {//updating voltage to alter velocity of the motor
+    leaderFlywheelMotor.setControl(
+        new VelocityVoltage(
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java:78]: I'd create a `VelocityVoltage` `voltageRequest` & instantiate within the constructor so you aren't creating a new `VelocityVoltage` every time you set the control.

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  @Override
+  public void stop() { //stops the motor
+    leaderFlywheelMotor.stopMotor();
+  }
+
+  @Override
+  public void configurePID(double kP, double kI, double kD) { //sets the pid values
+    var config = new Slot0Configs();
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOTalonFX.java:96]: IMO, you guys should configure PID & FF within this implementation, as we have followed that design schema throughout the entire repo. In the future we may reevaluate, but for now id just configure it here.

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java (hunk 4 above)
Comment (JacksonElia) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:33]: 12 volts

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java (hunk 5 above)
Comment (JacksonElia) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:55]: 60

Comment (JacksonElia) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:55]: or at least a comment saying converting to rps

Comment (JacksonElia) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:55]: ideally we would have a conversions thing (maybe in utils, or just in constants) that make it easy and clear whenever doing stuff like this

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIO.java (hunk 9 above)
Comment (JacksonElia) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIO.java:21]: and put spaces after those //

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java (hunk 5 above)
Comment (Ishan1522) [reply src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:55]: i think add a comment for now and we can add a conversion thing later

**Comment by aridavidson001** [comment]
if this is shooter code should the roller be here as well? or is this just for the flywheels?

**Comment by Ishan1522** [comment]
at some point id like to separate out the rollers and stuff, but for now the roller on the shooter should go in the flywheel subsystem

**Comment by Ishan1522** [comment]
U also need a DigitalInput for the beam break

**Comment by aridavidson001** [comment]
> at some point id like to separate out the rollers and stuff, but for now the roller on the shooter should go in the flywheel subsystem

wait what's the indexer then? cause I just realized you wrote that code.

**Comment by Ishan1522** [comment]
image
indexer = the one on the bottom
rollers = attached to the pivot

File: src/main/java/frc/robot/Constants.java [hunk 16]
```diff
@@ -156,8 +156,30 @@ public static final class ModuleConstants {
     public static final double DRIVE_A = 0.32 * WHEEL_CIRCUMFERENCE_METERS / DRIVE_GEAR_RATIO; // = 0.02225 V*s^2/m
   }
 
-  public static final class VisionConstants {
+  public static final class FlywheelConstants {
```
Comment (Ishan1522) [inline src/main/java/frc/robot/Constants.java:159]: flywheel constants should be moved to subsystems/shooter

**Comment by Ishan1522** [comment]
pull from main and format

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+  private FlywheelSim sim = new FlywheelSim(DCMotor.getKrakenX60(2), FlywheelConstants.GEAR_RATIO, 0.004); //initiates virtual
+  private DIOSim noteSensorSim = new DIOSim(new DigitalInput(FlywheelConstants.NOTE_SENSOR_ID));
+  private PIDController pid = new PIDController(FlywheelConstants.FLYWHEEL_P, FlywheelConstants.FLYWHEEL_I, FlywheelConstants.FLYWHEEL_D);
+  private SimpleMotorFeedforward feedFoward = new SimpleMotorFeedforward(FlywheelConstants.FLYWHEEL_S, FlywheelConstants.FLYWHEEL_V, FlywheelConstants.FLYWHEEL_A);
+  
+  private double appliedVolts = 0.0;
+
+  @Override
+  public void updateInputs(FlywheelIOInputs inputs) {
+    appliedVolts =
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:38]: i dont rlly get the point of this, u r already doing pid in the `setVelocity` method and setting the voltage in your `setVoltage` method.

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+  public void setVoltage(double volts) {
+    appliedVolts = volts;
+    sim.setInputVoltage(volts);
+  }
+/**
+ * @param velocityRPM User inputs the desired velocity in RPM, gets converted in method for PID to set value in RPS 
+ */
+  @Override
+  public void setVelocity(double velocityRPM) { //ffvolts is feedorward
+    sim.setState(pid.calculate(velocityRPM / 60.0));
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:61]: add feedforward

File: src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+import edu.wpi.first.math.system.plant.DCMotor;
+import edu.wpi.first.units.Unit;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.simulation.DIOSim;
+import edu.wpi.first.wpilibj.simulation.FlywheelSim;
+import frc.robot.Constants.FlywheelConstants;
+
+public class FlywheelIOSim implements FlywheelIO { //FlywheelIOSim makes Advantage kit log simulated movements using physics 
+  private FlywheelSim sim = new FlywheelSim(DCMotor.getKrakenX60(2), FlywheelConstants.GEAR_RATIO, 0.004); //initiates virtual
```
Comment (Ishan1522) [inline src/main/java/frc/robot/subsystems/shooter/FlywheelIOSim.java:29]: rename this to flywheelSim, its more explicit.

**Comment by Ishan1522** [comment]
yall dont mark shit as resolved if it isnt resolved
